HitScoreVisualizer/Config-Documentation.md
## HitScoreVisualizer config documentation

Below is a series of descriptions that relate to each value in a HitScoreVisualizer json config

### "majorVersion" "minorVersion" "patchVersion"
If the version number (excluding patch version) of the config is higher than that of the plugin, the config will not be loaded. If the version number of the config is lower than that of the plugin, the file will be automatically converted. Conversion is not guaranteed to occur, or be accurate, across major versions

### "displayMode"
- If not set, by default, it will display judgment text above score.
- If set to "scoreOnTop", it is like default but with score above.
- If set to "format", displays the judgment text, with the following format specifiers allowed:
  - %b: The score contributed by the part of the swing before cutting the block
  - %c: The score contributed by the accuracy of the cut
  - %a: The score contributed by the part of the swing after cutting the block
  - %t: The time dependence of the swing
  - %B, %C, %A, %T: As above, except using the appropriate judgment from that part of the swing (as configured for "beforeCutAngleJudgments", "accuracyJudgments", "afterCutAngleJudgments", or "timeDependencyJudgments")
  - %d: An arrow that points towards the center line of the note relative to the cut line
  - %s: The total score for the cut
  - %p: The percent out of 115 you achieved with your swing's score
  - %%: A literal percent symbol[README.md](../README.md)
  - %n: A newline
- If set to "textOnly", displays only the judgment text.
- If set to "numeric", displays only the note score.
- If set to "directions", displays judgement text and off-direction arrow.

### "fixedPosition"
If not null, judgments will appear and stay at rather than moving as normal, this will take priority over TargetPositionOffset. Additionally, the previous judgment will disappear when a new one is created (so there won't be overlap)

### "targetPositionOffset"
Will offset the target position of the hitscore fade animation. If a fixed position is defined in the config, that one will take priority over this one and this will be fully ignored

### "timeDependencyDecimalPrecision"
Number of decimal places to show time dependence to

### "timeDependencyDecimalOffset"
Which power of 10 to multiply the time dependence by

### "doIntermediateUpdates"
Disabling this only updates the score number once the note is cut and once the post swing score is done calculating; this slightly improves performance

### "assumeMaxPostSwing"
When the note is first cut, should the post swing score show the max rating before it finishes calculating

### "judgments"
Order from highest threshold to lowest; the first matching judgment will be applied

### "chainHeadJudgments"
Same as normal judgments but for burst sliders aka. chain notes

### "chainLinkDisplay"
Text displayed for burst slider segments

### "beforeCutAngleJudgments"
Judgments for the part of the swing before cutting the block (score is from 0-70)

### "accuracyJudgments"
Judgments for the accuracy of the cut (how close to the center of the block the cut was, score is from 0-15)

### "afterCutAngleJudgments"
Judgments for the part of the swing after cutting the block (score is from 0-30)

### "timeDependencyJudgments"
Judgments for time dependence (score is from 0-1)

### "badCutDisplays"
What to show when achieving a bad cut, either through hitting a note in the wrong direction or with the wrong color saber, or by hitting a bomb

### "randomizeBadCutDisplays"
Whether to randomize the order in which bad cut displays appear

### "missDisplays"
What to show missing a note

### "randomizeMissDisplays"
Whether to randomize the order in which miss displays appear

### Example Config
```json
{
  "majorVersion": 3,
  "minorVersion": 6,
  "patchVersion": 0,
  "displayMode": "format",
  "fixedPosition": null,
  "targetPositionOffset": null,
  "timeDependencyDecimalPrecision": 1,
  "timeDependencyDecimalOffset": 2,
  "doIntermediateUpdates": true,
  "assumeMaxPostSwing": false,
  "judgments": [
    { "threshold": 115, "text": "<size=250%><b>•", "color": [1, 1, 1, 1] },
    { "threshold": 108, "text": "<size=120%>%B%c%A", "color": [1, 1, 1, 1] },
    { "threshold": 0, "text": "<size=120%><alpha=#88>%B%c%A", "color": [1, 1, 1, 1] }
  ],
  "chainHeadJudgments":[
    { "threshold": 85, "text": "<size=250%><b>•", "color": [1, 1, 1, 1] },
    { "threshold": 78, "text": "<size=120%>%B%c", "color": [1, 1, 1, 1] },
    { "threshold": 0, "text": "<size=120%><alpha=#88>%B%c", "color": [1, 1, 1, 1] }
  ],
  "chainLinkDisplay": {
    "text": "<alpha=#66>•",
    "color": [1, 1, 1, 1]
  },
  "beforeCutAngleJudgments": [
    { "threshold": 70, "text": " " },
    { "threshold": 63, "text": "-" },
    { "threshold": 56, "text": "<color=#FFCC44>-</color>" },
    { "threshold": 0, "text": "<color=#FF0000>-</color>" }
  ],
  "accuracyJudgments": [],
  "afterCutAngleJudgments": [
    { "threshold": 30, "text": " " },
    { "threshold": 27, "text": "-" },
    { "threshold": 24, "text": "<color=#FFCC44>-</color>" },
    { "threshold": 0, "text": "<color=#FF0000>-</color>" }
  ],
  "timeDependencyJudgments": [],
  "randomizeBadCutDisplays": false,
  "badCutDisplays": [
	  { "text": "Bad Cut", "type": "WrongDirection", "color": [1, 1, 1, 1] },
	  { "text": "Bad Cut", "type": "WrongColor", "color": [1, 1, 1, 1] },
	  { "text": "Bomb", "type": "Bomb", "color": [1, 1, 1, 1] }
  ],
  "randomizeMissDisplays": true,
  "missDisplays": [
	  { "text": "MISS", "color": [1, 1, 1, 1] },
	  { "text": "MOSS", "color": [1, 1, 1, 1] },
	  { "text": "MASS", "color": [1, 1, 1, 1] },
	  { "text": "MESS", "color": [1, 1, 1, 1] }
  ]
}
```

changelog.md
## Changelog

### 3.6.1

- Added a button to open the config folder in the file explorer
- Fixed judgment thresholds being falsely invalidated as having "duplicate thresholds"

### 3.6.0

- Thresholds now must be sorted correctly in descending order in order for configs to be compatible
- Removed `isDefaultConfig` - you don't need to update your configs, it will just be ignored
- Added customization for miss and bad cut effects with custom text

```json
"badCutDisplays": [
	{
		"text": "A Cut Of The Bad Variety",
		"color": [1.0, 0.7, 0.7, 1.0]
	}
],
"missDisplays": [
	{
		"text": "You Should Work On Your Aim",
		"color": [0.6, 1.0, 0.7, 1.0]
	}
]
```

- Miss and bad cut effects will be randomized by default unless you specify `randomizeBadCutDisplays` and `randomizeMissDisplays`

```json
"randomizeBadCutDisplays": false,
"randomizeMissDisplays": false
```

- For bad cuts, you can specify which types of bad cuts the display will show for. The options are `WrongDirection`, `WrongColor`, and `Bomb`. Not specifying a type will make the display show for all
types of bad cuts

```json
"badCutDisplays": [
	{
		"text": "<size=600%>BOOM",
		"type": "Bomb",
		"color": [1.0, 1.0, 1.0, 1.0]
	}
]
```

- HitScoreVisualizer will recreate the default config file each time the game is loaded
- Updated the default config to be more like the default config for the HitScoreVisualizer quest mod

### 3.5.3
- Added spaces to the main menu button

### 3.5.2
- Fixed bloom font causing black screen when Enhancements was not installed

### 3.5.1
- Added a setting to allow hit score text to be shown while "No Texts and Huds" is enabled
- Fixed bloom font showing over notes and environment objects
- Fixed "assumeMaxPostSwing" not adding the full post swing score to the total score, which was causing lower judgment thresholds to be used
- Moved settings to the left screen and tweaked some interface elements' scale

### 3.5.0
- Added a toggle in configs that will decide, when the note is first cut, if the score display will show the max possible after cut score
```json
"assumeMaxPostSwing": true
```
Note: it is recommended to set `doIntermediateUpdates` to `false` in order for this to be noticeable. This setting is generally only useful if you're swinging extremely slow, like on "poodle"
maps, for instance.
- Added the `%d` token which, when used in `format` display mode, will be replaced with an arrow that points in the direction relative to the cut line towards the center line of the note
```json
"judgments": [
    { "threshold": 115, "text": "%d%n%s", "color": [1, 1, 1, 1] },
]
```

### 3.4.0
- Added a button to toggle score effect's italics
- Added judgments for chain head notes. These work the way that normal judgments work but are only applied when a chain is cut. Example below
```json
"chainHeadJudgments":[
  { "threshold": 85, "text": "<size=250%><b>•", "color": [1, 1, 1, 1] },
  { "threshold": 78, "text": "<size=120%>%B%c", "color": [1, 1, 1, 1] },
  { "threshold": 0, "text": "<size=120%><alpha=#88>%B%c", "color": [1, 1, 1, 1] }
]
```
- Added configuration of chain segment (aka. chain links) display. Example below
```json
"chainLinkDisplay": {
  "text": "<alpha=#66>%s",
  "color": [1, 1, 1, 1]
}
```
- The mod will now only attempt to load `.json` files in the configs folder
- Some minor optimizations

### 3.3.8
- Fixed an error causing hitscore bloom breaking the game on level load

### 3.3.7
- Fixed arc notes inconsistently showing hit score

HitScoreVisualizer/Settings/Configuration.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using HitScoreVisualizer.Utilities.Json;
using Newtonsoft.Json;
using UnityEngine;
using Version = Hive.Versioning.Version;

namespace HitScoreVisualizer.Settings;

public class Configuration
{
	[JsonIgnore]
	internal static Configuration Default { get; } = new Configuration
	{
		Version = Plugin.Metadata.HVersion,
		IsDefaultConfig = true,
		DisplayMode = "format",
		DoIntermediateUpdates = true,
		AssumeMaxPostSwing = false,
		TimeDependenceDecimalPrecision = 1,
		TimeDependenceDecimalOffset = 2,
		NormalJudgments =
		[
			new(115, "%BFantastic%A%n%s", [1.0f, 1.0f, 1.0f, 1.0f]),
			new(101, "<size=80%>%BExcellent%A</size>%n%s", [0.0f, 1.0f, 0.0f, 1.0f]),
			new(90, "<size=80%>%BGreat%A</size>%n%s", [1.0f, 0.980392158f, 0.0f, 1.0f]),
			new(80, "<size=80%>%BGood%A</size>%n%s", [1.0f, 0.6f, 0.0f, 1.0f], true),
			new(60, "<size=80%>%BDecent%A</size>%n%s", [1.0f, 0.0f, 0.0f, 1.0f], true),
			new(0, "<size=80%>%BWay Off%A</size>%n%s", [0.5f, 0.0f, 0.0f, 1.0f], true)
		],
		ChainHeadJudgments =
		[
			new(115, "%BFantastic%n%s", [1.0f, 1.0f, 1.0f, 1.0f]),
			new(101, "<size=80%>%BExcellent</size>%n%s", [0.0f, 1.0f, 0.0f, 1.0f]),
			new(90, "<size=80%>%BGreat</size>%n%s", [1.0f, 0.980392158f, 0.0f, 1.0f]),
			new(80, "<size=80%>%BGood</size>%n%s", [1.0f, 0.6f, 0.0f, 1.0f], true),
			new(60, "<size=80%>%BDecent</size>%n%s", [1.0f, 0.0f, 0.0f, 1.0f], true),
			new(0, "<size=80%>%BWay Off</size>%n%s", [0.5f, 0.0f, 0.0f, 1.0f], true)
		],
		BeforeCutAngleJudgments =
		[
			new(70, " + "),
			new(0, " ")
		],
		AccuracyJudgments =
		[
			new(15, " + "),
			new(0, " ")
		],
		AfterCutAngleJudgments =
		[
			new(30, " + "),
			new(0, " ")
		]
	};

	[JsonProperty("majorVersion", DefaultValueHandling = DefaultValueHandling.Include)]
	public ulong MajorVersion { get; private set; } = Plugin.Metadata.HVersion.Major;

	[JsonProperty("minorVersion", DefaultValueHandling = DefaultValueHandling.Include)]
	public ulong MinorVersion { get; private set; } = Plugin.Metadata.HVersion.Minor;

	[JsonProperty("patchVersion", DefaultValueHandling = DefaultValueHandling.Include)]
	public ulong PatchVersion { get; private set; } = Plugin.Metadata.HVersion.Patch;

	[JsonIgnore]
	internal Version Version
	{
		get => new(MajorVersion, MinorVersion, PatchVersion);
		set
		{
			MajorVersion = value.Major;
			MinorVersion = value.Minor;
			PatchVersion = value.Patch;
		}
	}

	[JsonProperty("isDefaultConfig")]
	public bool IsDefaultConfig { get; internal set; }

	[JsonProperty("displayMode")]
	[DefaultValue("")]
	public string DisplayMode { get; internal set; } = string.Empty;

	[JsonProperty("useFixedPos")]
	[DefaultValue(false)]
	[ShouldNotSerialize]
	[Obsolete("Obsolete since 3.2.0. Either use the FixedPosition property or leave as null instead.")]
	public bool UseFixedPos { get; internal set; }

	[JsonProperty("fixedPosX")]
	[DefaultValue(0f)]
	[ShouldNotSerialize]
	[Obsolete("Obsolete since 3.2.0. Use the FixedPosition property instead.")]
	public float FixedPosX { get; internal set; }

	[JsonProperty("fixedPosY")]
	[DefaultValue(0f)]
	[ShouldNotSerialize]
	[Obsolete("Obsolete since 3.2.0. Use the FixedPosition property instead.")]
	public float FixedPosY { get; internal set; }

	[JsonProperty("fixedPosZ")]
	[DefaultValue(0f)]
	[ShouldNotSerialize]
	[Obsolete("Obsolete since 3.2.0. Use the FixedPosition property instead.")]
	public float FixedPosZ { get; internal set; }

	[JsonProperty("fixedPosition", NullValueHandling = NullValueHandling.Ignore)]
	public Vector3? FixedPosition { get; set; }

	[JsonProperty("targetPositionOffset", NullValueHandling = NullValueHandling.Ignore)]
	public Vector3? TargetPositionOffset { get; set; }

	[JsonProperty("doIntermediateUpdates")]
	public bool DoIntermediateUpdates { get; internal set; }

	[JsonProperty("assumeMaxPostSwing")]
	public bool AssumeMaxPostSwing { get; internal set; }

	[JsonProperty("timeDependencyDecimalPrecision")]
	[DefaultValue(1)]
	public int TimeDependenceDecimalPrecision { get; internal set; }

	[JsonProperty("timeDependencyDecimalOffset")]
	[DefaultValue(2)]
	public int TimeDependenceDecimalOffset { get; internal set; }

	[JsonProperty("judgments")]
	public List<NormalJudgment>? NormalJudgments { get; internal set; }

	[JsonProperty("chainHeadJudgments")]
	public List<ChainHeadJudgment>? ChainHeadJudgments { get; internal set; }

	[JsonProperty("chainLinkDisplay")]
	public ChainLinkDisplay? ChainLinkDisplay { get; internal set; }

	[JsonProperty("beforeCutAngleJudgments")]
	public List<JudgmentSegment>? BeforeCutAngleJudgments { get; internal set; }

	[JsonProperty("accuracyJudgments")]
	public List<JudgmentSegment>? AccuracyJudgments { get; internal set; }

	[JsonProperty("afterCutAngleJudgments")]
	public List<JudgmentSegment>? AfterCutAngleJudgments { get; internal set; }

	[JsonProperty("timeDependencyJudgments")]
	public List<TimeDependenceJudgmentSegment>? TimeDependenceJudgments { get; internal set; }
}